Support NXP ENET_1G gigabit ethernet on mimxrt1170_evk

stanislav-poboril opened the pull request:
This PR addresses issue #66348 on mimxrt1170_evk board.

- Updated the experimental eth_nxp_enet driver, created by @decsny, to support RGMII mode and ENET_1G peripheral.
- Added support for Realtek RTL8211F ethernet PHY which is connected to ENET_1G on a mentioned board.
- Performed necessary updates in platform SoC initialization and clock control code.
- Since it uses the experimental eth_nxp_enet driver, the overlay file and documentation on its use were added.

Along with this change, the Microchip KSZ8081 PHY driver needed to be updated a little, so the eth_nxp_enet can work with both PHYs. Also fixed few issues found in that PHY driver when doing the changes.

stanislav-poboril committed:
drivers: clock_control: mcux_ccm_rev2: Add ENET_1G clock

Add ENET_1G clock value to the RT11XX CCM version.
Implemented enabling ENET_1G clock and getting its frequency.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
soc: rt11xx: support nxp_enet in soc

Support NXP ENET_1G on RT11xx soc.

Fixes: #66348

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
dts: bindings: ethernet-controller: Add RGMII mode

Add option for RGMII connection type between the MAC and the PHY
device into the ethernet controller binding.

Fixes: #66348

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
dts: bindings: Add RGMII mode definition to NXP ENET bindings

Add RGMII mode definition to be used in bindings related to NXP ENET_1G IP.

Fixes: #66348

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
drivers: nxp_enet: Support RGMII mode for ENET_1G

- Added nxp,enet1g compatible to distinguish between ENET (nxp,enet)
and ENET_1G (nxp,enet1g) peripherals within the same driver.
- Added config ETH_NXP_ENET_1G to enable 1G mode of operation on ENET_1G.
- Support RGMII mode of connection between MDIO and PHY to be
able to work with ENET_1G peripheral and support 1000M speed.
- Removed performing of PHY reset before configuring link - it is
not desirable for RTL8211F PHY connected to ENET_1G on RT1170.
Reset of other PHYs can be performed by PHY driver itself if required.

Fixes: #66348

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
drivers: ethernet: phy: Add Realtek RTL8211F PHY driver

Add driver for Realtek RTL8211F 10/100/1000M ethernet PHY.
This driver implements vendor specific behaviour like
detecting link state change by GPIO interrupt, which is not
present in the generic MII driver.

Fixes: #66348

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
boards: nxp: Support NXP ENET_1G on mimxrt1170_evk

Support NXP ENET_1G on mimxrt1170_evk/mimxrt1176/cm7 platform.
Added test configuration sample.net.zperf.nxp_enet1g and
documented the usage of the ethernet driver with ENET_1G
peripheral.

Fixes: #66348

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

decsny commented:
@stanislav-poboril trust in the zephyr-bot

decsny commented on include/zephyr/dt-bindings/clock/imx_ccm_rev2.h:
```diff
@@ -92,8 +92,13 @@
 #define IMX_CCM_ENET_CLK	0x3000UL
 #define IMX_CCM_ENET_PLL	0x3001UL
 
+
```
might have added this newline by accident

decsny commented on soc/arm/nxp_imx/rt/soc_rt11xx.c:
```diff
@@ -414,22 +414,15 @@ static ALWAYS_INLINE void clock_init(void)
-#if CONFIG_ETH_MCUX_RMII_EXT_CLK
+	/* Set ENET1G TX_CLK to be driven by ENET2_CLK_ROOT and output on TX_CLK_IO pad */
+	IOMUXC_GPR->GPR5 = (IOMUXC_GPR_GPR5_ENET1G_RGMII_EN(0x01U) |
+		(IOMUXC_GPR->GPR5 & ~IOMUXC_GPR_GPR5_ENET1G_TX_CLK_SEL(0x01U)));
```
since this is in SOC level, should we check the phy connection type property to determine the clocking ? maybe the enet1G is not always used with RGMII ?

stanislav-poboril replied:
Updated, if it is not rgmii, the original code is called.

stanislav-poboril commented:
Are the errors:

```
FileNotFoundError: [Errno 2] No such file or directory: '_test_plan_partial.json'
```
```
FileNotFoundError: [Errno 2] No such file or directory: 'artifacts/*/*/twister.xml'
```

indicating issues with automation or is there something wrong with my PR?

Thanks.

decsny commented:
> Are the errors:
> 
> ```
> FileNotFoundError: [Errno 2] No such file or directory: '_test_plan_partial.json'
> ```
> ```
> FileNotFoundError: [Errno 2] No such file or directory: 'artifacts/*/*/twister.xml'
> ```
> 
> indicating issues with automation or is there something wrong with my PR?
> 
> Thanks.

Something went wrong with the CI I just restarted it

decsny commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -308,6 +315,14 @@ static enum ethernet_hw_caps eth_nxp_enet_get_capabilities(const struct device *
+	if (config->phy_mode == NXP_ENET_RGMII_MODE) {
+		caps |= ETHERNET_LINK_1000BASE_T;
+	}
+#endif /* ENET_HAS_AVB */
```
why is it guarded by ENET_HAS_AVB?

stanislav-poboril replied:
`ENET_HAS_AVB` is defined if `defined(FSL_FEATURE_ENET_HAS_AVB) && FSL_FEATURE_ENET_HAS_AVB`.
`FSL_FEATURE_ENET_HAS_AVB` is used by the underlying SDK ENET driver to determine if it should define  `kENET_RgmiiMode`, `kENET_MiiSpeed1000M` etc.

It could use `#if defined(ENET_1G)` instead, but I am using it here for consistency with the SDK driver.
According to user manual, the presence of Audio Video Bridging implies ENET_1G:

```
ENET supports 10M/100M Ethernet, and has the following key features:
• 3.3V RMII and MII operation
• IEEE1588

ENET_1G supports 1G Ethernet, and has the following features:
• RGMII/RMII/MII operation
• IEEE1588
• AVB
```

decsny replied:
There should probably be a kconfig indicating ENET1G support, and we would use that instead

stanislav-poboril replied:
I will try to improve this and add KConfig option.

stanislav-poboril replied:
Hi @decsny,

So I was thinking about adding `HAS_MCUX_ENET1G` into `modules/Kconfig.mcux` as there is also `HAS_MCUX_ENET` (which I am not sure if it is used anywhere).
Then there is `ETH_NXP_ENET` in `drivers/ethernet/Kconfig.nxp_enet`. `CONFIG_ETH_NXP_ENET` is used from clock control and SoC code. I am not sure if adding `ETH_NXP_ENET1G` is ok, as both share the same driver.
Then there are definitions like `config ETH_NXP_ENET_HW_ACCELERATION`.

So we have about four options:

- Using `CONFIG_HAS_MCUX_ENET1G`.
- Using `CONFIG_ETH_NXP_ENET1G`.
- Using `CONFIG_ETH_NXP_ENET_ENET1G` (or similar name) as a subconfig.
- Or asking device tree if `ENET1G` is enabled.

Which one is preferred? Or something different? Thanks.

decsny replied:
Should not add HAS_MCUX_ENET1G, those kconfigs in that file we are trying to deprecate, as I think they were from before we used devicetree to determine if a platform has a device.

I think the name of the kconfig does not matter so much, but my first instinct is that it should be a dependency of the driver kconfig. ie the kconfig file for the driver would look basically like:
```
config ETH_NXP_ENET
...

if ETH_NXP_ENET

config ETH_NXP_ENET1G (or whatever name)
bool
default y
depends on DT_HAS_NXP_ENET_1G_ENABLED
....

endif
```

and therefore yes, I am thinking it might be correct to add another compatible to DT for this purpose. i.e., the node would have 2 compatibles, "nxp,enet" and "nxp,enet1g". Same driver but the 2nd compatible indicates the variation of the device. I will admit I haven't actually done this before but I would like to experiment with this idea to see how it works, you will probably need to add like an empty binding so that the DT tooling makes the kconfig and macros for it.

The reason I say this is because then the code can be conditionally included without user needing to manually select the kconfig, DT is basically THE standard way for us to automatically configure the software based on the platform needs, so it makes sense to me to attempt this sort of approach.

decsny replied:
and also, I think you could even add potentially a static assertion (using BUILD_ASSERT macro) in this driver device init macro that checks that in the case if the phy has RGMII/1000-Base_T, but does not have the enet1g compatible, then an error is raised

decsny commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -509,19 +521,44 @@ static void nxp_enet_phy_cb(const struct device *phy,
+			speed = kENET_MiiSpeed1000M;
+		} else if (PHY_LINK_IS_SPEED_100M(state->speed)) {
+#else
+		if (PHY_LINK_IS_SPEED_100M(state->speed)) {
```
nit but I think compliance allows you to do

```
} else
#endif
if (...) {
```

stanislav-poboril replied:
Checkpatch does not take that (suspect code indent for conditional statements).

decsny replied:
well, that's fine, but this comment would be irrelevant if using a kconfig, since we can use the IS_ENABLED macro then and avoid so much conditional compilation in the driver (my goal is to avoid preprocessor #ifdefs as much as possible)

jukkar replied:
I agree, best is to use `IS_ENABLED()` or different variants if possible.

decsny commented on dts/bindings/ethernet/realtek,rtl8211f.yaml:
```diff
@@ -0,0 +1,18 @@
+    type: phandle-array
+    specifier-space: gpio
+    description: GPIO connected to PHY reset signal pin. Reset is active low.
+  rt,interrupt-gpio:
```
since these properties are on other phys, I wonder if we can generalize them (not required, but a suggestion)

decsny commented:
@stanislav-poboril the ci runner seems to be having problem doing a git rebase, there's some conflicts, I think you will need to rebase the branch

dleach02 commented:
@stanislav-poboril, please rebase to clear the conflicts

decsny commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
I just cherry picked the commit to the PR deprecating eth_mcux, so resolved I guess

stanislav-poboril replied:
So if the other PR is planned to be merged first, I will remove it from here (or let it stay here as hopefully having an already merged commit here will not create a merge conflict).

decsny commented on boards/nxp/mimxrt1170_evk/doc/index.rst:
can you also update the feature table at the top of this document to indicate the support of the ENET1G (and please also remove the mention of the enet_qos while youre at it, someone seemed to think it was the same thing maybe)

stanislav-poboril replied:
I have added ENET_QOS on another line and marked as not supported.

stanislav-poboril commented on soc/nxp/imxrt/imxrt11xx/soc.c:
```diff
@@ -414,6 +414,17 @@ static ALWAYS_INLINE void clock_init(void)
 #endif
 #endif
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(enet1g), okay)
+#if DT_ENUM_HAS_VALUE(DT_CHILD(DT_NODELABEL(enet1g), ethernet), phy_connection_type, rgmii)
```
Actually each ifdef branch does something else, I reduced just one duplicate line.

stanislav-poboril commented:
> @stanislav-poboril the ci runner seems to be having problem doing a git rebase, there's some conflicts, I think you will need to rebase the branch

@decsny I have rebased but there are still errors like:

```
Error: Error: failed to run script step: command terminated with non-zero exit code: Error executing in Docker Container: 1
Error: Process completed with exit code 1.
Error: Executing the custom container implementation failed. Please contact your self hosted runner administrator.
```

jukkar commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -672,6 +711,10 @@ static int eth_nxp_enet_init(const struct device *dev)
 		enet_config.miiMode = kENET_MiiMode;
 	} else if (config->phy_mode == NXP_ENET_RMII_MODE) {
 		enet_config.miiMode = kENET_RmiiMode;
+#ifdef ENET_HAS_AVB
+	} else if (config->phy_mode == NXP_ENET_RGMII_MODE) {
+		enet_config.miiMode = kENET_RgmiiMode;
+#endif /* ENET_HAS_AVB */
```
You could do this in order to avoid using `#ifdef`
```
} else if (COND_CODE_1(ENET_HAS_AVB,
	   (config->phy_mode == NXP_ENET_RGMII_MODE), (0))) {
	enet_config.miiMode = kENET_RgmiiMode;
}
```

stanislav-poboril replied:
Used when possible.

decsny commented:
> > @stanislav-poboril the ci runner seems to be having problem doing a git rebase, there's some conflicts, I think you will need to rebase the branch
> 
> @decsny I have rebased but there are still errors like:
> 
> ```
> Error: Error: failed to run script step: command terminated with non-zero exit code: Error executing in Docker Container: 1
> Error: Process completed with exit code 1.
> Error: Executing the custom container implementation failed. Please contact your self hosted runner administrator.
> ```

I restarted it again

decsny commented:
@stanislav-poboril there are some conflicts again

stanislav-poboril commented:
> @stanislav-poboril there are some conflicts again

- rebased to upstream's main branch
- removed the commit with PHYKSZ8081 driver updates already present on upstream
- since eth_nxp_enet.c driver is no longer experimental:
    - moved also the device tree and pinctrl updates from enet1g overlay to a regular location
    - removed the experimental mentions in docs and names
    - nxp,enet1g.overlay reduced to only switch enet1g related components on
- TODO: not replaced the using of the FSL_FEATURE_ENET_HAS_AVB macro by Kconfig yet.

stanislav-poboril commented:
* Rebased to upstream's main branch.
* Introduced new nxp,enet1g compatible to distinguish between ENET (nxp,enet) and ENET_1G (nxp,enet1g) peripherals within the same driver.
* Introduced new Kconfig ETH_NXP_ENET_1G to enable 1G mode of operation on ENET_1G.
* Replaced the using of the FSL_FEATURE_ENET_HAS_AVB macro by ETH_NXP_ENET_1G.
* Added compile time check: if RGMII phy connection mode is selected, nxp,enet1g compatible must be used too.

stanislav-poboril commented on boards/nxp/mimxrt1170_evk/doc/index.rst:
```diff
@@ -144,8 +144,11 @@ RT1170 EVKB (`mimxrt1170_evk@B//cm7/cm4`)
++-----------+------------+-------------------------------------+-----------------+-----------------+
+| ENET1G    | on-chip    | ethernet - 10/100/1000M             | Supported (M7)  | No support      |
++-----------+------------+-------------------------------------+-----------------+-----------------+
+| ENET_QOS  | on-chip    | ethernet quality of service module  | No support      | No support      |
```
Removed.

decsny commented on drivers/ethernet/nxp_enet/eth_nxp_enet.c:
```diff
@@ -908,3 +955,23 @@ int nxp_enet_##n##_init(void)								\
+
+#define NXP_ENET1G_INIT(n)								\
+											\
+int nxp_enet1g_##n##_init(void)								\
```
This function looks the same as the one above, am I missing some difference?

If they are the same you can just call the DT for each macro twice, once for each compat, on the same function

stanislav-poboril replied:
One function will expand to extracting "nxp,enet" compatible nodes data and the other to extracting "nxp,enet1g" compatible nodes data. I am not sure if it will work when DT_DRV_COMPAT will be always set to nxp_enet.

And if there will be both "nxp,enet" and "nxp,enet1g" compatible nodes with ok status, both will have instance number 0, won't they?

decsny replied:
You can #undef the DT_DRV_COMPAT before it and re-#define it or just don't used the INST macro and specify the compatible as an argument

stanislav-poboril replied:
Done.

stanislav-poboril commented:
rebased

dleach02 commented:
@galak I've assigned this over to me as it is an NXP IP/driver. If you object then just assign it back.

dleach02 commented:
@stanislav-poboril please correct the CI issues

stanislav-poboril commented:
Updates in PHY RTL8211F:

- PHY dependency on GPIO is optional. If desired, it must be enabled by Kconfig (CONFIG_PHY_REALTEK_RTL8211F_GPIO_SUPPORT). Some platforms don't have GPIO available.
- Fixed PHY reset fallback code for resetting by register if GPIO reset pin is not set.
- Fixed handling of multiple PHY instances, where some may have GPIO set while others not. The CONFIG_PHY_REALTEK_RTL8211F_GPIO_SUPPORT controls if any instance can support GPIO reset or interrupt. Then the device tree properties rt,reset-gpio and rt,interrupt-gpio control, whether the particular PHY instance uses GPIO for reset or interrupt link state detection functions (or both of them or none of them).

stanislav-poboril committed:
boards: nxp: Adjusted device tree on vmu_rt1170 for updated ENET_1G.

Adjusted the board's device tree after ENET_1G compatible changed
from "nxp,kinetis-ethernet" to "nxp,enet1g".

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

pdgendt commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -41,13 +41,34 @@ config PHY_ADIN2111
+
+if PHY_REALTEK_RTL8211F
+
+config PHY_REALTEK_RTL8211F_GPIO_SUPPORT
```
I don't think you need a `Kconfig` for this, you can check using devicetree if a compatible has a `rst-gpios` property.

stanislav-poboril replied:
That's how it was before in `c` code, but I've changed this after @JiafeiPan asked me to remove the mandatory dependency on `GPIO` (build problems on other platform). The dependency is added via `Kconfig` and it looks like the preprocessor currently does not have a function to tell if any compatible node instance has the property.

decsny replied:
The reason the kconfig is needed is to conditionalize the dependency on GPIO to avoid build errors, the PR was already as you're describing originally and that didn't work

pdgendt replied:
The macro has made it easier since #70385

decsny replied:
The PR used this macro as of a few days ago, the problem is not with the code file, the problem is with the KConfig dependency on GPIO, if it is unconditional then platforms without GPIO cannot build this driver

pdgendt replied:
Platforms without GPIO wouldn't add the gpio properties to the device instance?

stanislav-poboril replied:
Yes, but how do you remove `depends on GPIO` from `Kconfig`?
That dependency would remove the driver sources from build.

pdgendt replied:
Where you do this now:
```c
#if CONFIG_PHY_REALTEK_RTL8211F_GPIO_SUPPORT
	if (gpio_is_ready_dt(&config->reset_gpio)) {
		// ....
	}
#endif /* CONFIG_PHY_REALTEK_RTL8211F_GPIO_SUPPORT */
```

do

```c
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(rt_reset_gpio)
	if (gpio_is_ready_dt(&config->reset_gpio)) {
		// ....
	}
#endif /* DT_ANY_INST_HAS_PROP_STATUS_OKAY(rt_reset_gpio) */
```

decsny replied:
Again, this is what the PR was a few days ago. We changed it for a reason

pdgendt replied:
And I still fail to see the difficulty, maybe a more complete solution would be:

```c
#if defined(CONFIG_GPIO) && DT_ANY_INST_HAS_PROP_STATUS_OKAY(rt_reset_gpio)
#define RST_GPIO_SUPPORT 1
#endif

#if RST_GPIO_SUPPORT
// do something with gpio
#endif /* RST_GPIO_SUPPORT */
```

And you can `imply GPIO` in your Kconfig if you would like to default enable it.

decsny replied:
> And I still fail to see the difficulty, maybe a more complete solution would be:
> 
> ```c
> #if defined(CONFIG_GPIO) && DT_ANY_INST_HAS_PROP_STATUS_OKAY(rt_reset_gpio)
> #define RST_GPIO_SUPPORT 1
> #endif
> 
> #if RST_GPIO_SUPPORT
> // do something with gpio
> #endif /* RST_GPIO_SUPPORT */
> ```

this isn't the issue, the issue is the `depends on GPIO` is needed sometimes and not others, and I don't think we want to reverse the dependency here, @JiafeiPan can explain better maybe

stanislav-poboril replied:
We cannot use `DT_ANY_INST_HAS_PROP_STATUS_OKAY` in `Kconfig` file to enable or disable `depends on GPIO`.
That's why there is the new redundant config item, to be able to bypass dependency on `GPIO` if it is not available on the platform.

pdgendt replied:
I won't block for this :) only a detail anyway.

Commit:
scripts: kconfig: Add dt_compat_any_has_prop

Add a kconfig preprocessor function to check if
any node of a certain compatible has a specific property in DT.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny reviewed (changes requested):
I just made  #73728 and tested it and it works, I tried adding and removing 'mc,reset-gpios' with the ksz8081 driver as an example, I think it can be used to solve this issue being discussed, let me know if you need help on how to use it

stanislav-poboril commented:
> I just made #73728 and tested it and it works, I tried adding and removing 'mc,reset-gpios' with the ksz8081 driver as an example, I think it can be used to solve this issue being discussed, let me know if you need help on how to use it

Cherry-picked commit from #73728, updated to use the new function and removed the redundant Kconfig item.
Updated vmu_rt1170 PHY to nxp,tja1103.

decsny reviewed (dismissed):
thanks for the patience on this PR @stanislav-poboril

stanislav-poboril commented:
Pushed additional updates for vmu_rt1170 after it has been reconfigured to use TJA1103 ethernet PHY.
(Legacy ethernet driver had hardcoded PHY KSZ8081, now the PHY  is configurable, so using the right one).

@bperseghetti, would it be possible to verify again if ethernet works on vmu_rt1170? Thank you.

decsny commented:
this compliance fail I've seen on a few PRs now where it doesn't seem related to the PR

decsny commented:
#73403 has a fix for the compliance issue

stanislav-poboril commented:
> #73403 has a fix for the compliance issue

Rebased to take it.

pdgendt commented on drivers/ethernet/phy/phy_realtek_rtl8211f.c:
```diff
@@ -0,0 +1,634 @@
+	ret = k_mutex_lock(&data->mutex, K_FOREVER);
+	if (ret) {
+		LOG_ERR("PHY mutex lock error");
+		goto done;
```
You shouldn't unlock if locking failed.

stanislav-poboril replied:
> You shouldn't unlock if locking failed.

Thanks, fixed.
